src/timeutils.cpp

```diff
@@ -0,0 +1,66 @@
+#include <time.h>
+
+#include "configcat/timeutils.h"
+
+// Even though gmtime_s is part of the C++17 according to the reference (https://en.cppreference.com/w/c/chrono/gmtime),
+// yet it's unavailable on some platforms. So, in those cases we need to "polyfill" it. This can be achieved using
+// the following SFINAE trick (see also https://slashvar.github.io/2019/08/17/Detecting-Functions-existence-with-SFINAE.html):
+
+template <typename T>
+auto gmtime_impl(T tm, const time_t* time, int) -> decltype(gmtime_s(tm, time)) {
+    return gmtime_s(tm, time);
+}
+
+template <typename T>
+struct tm* gmtime_impl(T tm, const time_t* time, float) {
+    auto pBuf = gmtime(time);
+    *tm = *pBuf;
+    return nullptr;
+}
+
+using namespace std;
+
+namespace configcat {
+
+// https://howardhinnant.github.io/date_algorithms.html#days_from_civil
+// From C++20, this can be replaced with std::chrono::sys_days
+template <class Int> constexpr Int days_from_civil(Int y, unsigned m, unsigned d) noexcept {
+    static_assert(std::numeric_limits<unsigned>::digits >= 18, "This algorithm has not been ported to a 16 bit unsigned integer");
+    static_assert(std::numeric_limits<Int>::digits >= 20, "This algorithm has not been ported to a 16 bit signed integer");
+    y -= m <= 2;
+    const Int era = (y >= 0 ? y : y-399) / 400;
+    const unsigned yoe = static_cast<unsigned>(y - era * 400);  // [0, 399]
+    const unsigned doy = (153*(m > 2 ? m-3 : m+9) + 2)/5 + d-1; // [0, 365]
+    const unsigned doe = yoe * 365 + yoe/4 - yoe/100 + doy;     // [0, 146096]
+    return era * 146097 + static_cast<Int>(doe) - 719468;
+}
```

Comment:
This is terrible. I mean, it's terrible that the standard library of a language in 2017 doesn't provide this out of the box... Dealing with datetimes is a very tricky business, so I guess it works but can't really verify it.

However, I don't really like those `static_assert`s. Although I don't know if the rest of the SDK can be compiled to 16-bit architectures, but exluding them because of a convenience helper method doesn't seem right.

Could we rather just exclude these helpers from 16-bit builds?